File: packages/autonavi_search_flutter/ios/Classes/AMapSearchAdapter.swift
import AMapSearchKit

/// Adapter that isolates all AMapSearchKit calls.
/// When upgrading the AMap Search SDK, only this file needs changes.
class AMapSearchAdapter: NSObject, AMapSearchDelegate {

    // MARK: - Types (no SDK types exposed)

    struct POIItem {
        let poiId: String?; let title: String?; let typeDes: String?; let typeCode: String?
        let latitude: Double?; let longitude: Double?; let address: String?; let tel: String?
        let distance: Int; let cityName: String?; let adName: String?
    }

    struct POIPage {
        let pois: [POIItem]; let totalCount: Int; let pageCount: Int; let pageNum: Int
    }

    struct AddressComponent {
        let formattedAddress: String?; let country: String?; let province: String?
        let city: String?; let cityCode: String?; let district: String?; let adCode: String?
        let street: String?; let streetNumber: String?; let township: String?; let townCode: String?
    }

    struct GeocodeItem {
        let formattedAddress: String?; let country: String?; let province: String?
        let city: String?; let cityCode: String?; let district: String?; let adCode: String?
        let latitude: Double?; let longitude: Double?; let level: String?
    }

    struct RoutePath {
        let distance: Int; let duration: Int; let strategy: String
        let tolls: Double; let tollDistance: Int; let trafficLights: Int
        let steps: [RouteStep]
    }

    struct RouteStep {
        let instruction: String?; let road: String?; let distance: Int; let duration: Int
        let action: String?; let path: [[String: Double]]
    }

    struct DistrictItem {
        let name: String?; let adCode: String?; let cityCode: String?; let level: String?
        let latitude: Double?; let longitude: Double?; let districts: [DistrictItem]
    }

    typealias POICallback       = (POIPage?, (code: String, message: String)?) -> Void
    typealias RegeocodeCallback = (AddressComponent?, (code: String, message: String)?) -> Void
    typealias GeocodeCallback   = ([GeocodeItem]?, (code: String, message: String)?) -> Void
    typealias RouteCallback     = ([RoutePath]?, (code: String, message: String)?) -> Void
    typealias DistrictCallback  = ([DistrictItem]?, (code: String, message: String)?) -> Void

    // MARK: - Private

    private var searchAPI: AMapSearchAPI?
    private var pendingPOI:      [ObjectIdentifier: (request: AMapPOISearchBaseRequest, callback: POICallback)] = [:]
    private var pendingRegeo:    [ObjectIdentifier: RegeocodeCallback] = [:]
    private var pendingGeo:      [ObjectIdentifier: GeocodeCallback] = [:]
    private var pendingRoute:    [ObjectIdentifier: RouteCallback] = [:]
    private var pendingDistrict: [ObjectIdentifier: DistrictCallback] = [:]

    override init() {
        super.init()
        searchAPI = AMapSearchAPI()
        searchAPI?.delegate = self
    }

    // MARK: - Public API

    func searchKeyword(keyword: String, city: String, types: String?,
                       pageSize: Int, page: Int, callback: @escaping POICallback) {
        let req = AMapPOIKeywordsSearchRequest()
        req.keywords = keyword; req.city = city; req.types = types
        req.offset = pageSize; req.page = page
        pendingPOI[ObjectIdentifier(req)] = (req, callback)
        searchAPI?.aMapPOIKeywordsSearch(req)
    }

    func searchNearby(latitude: Double, longitude: Double, radius: Int,
                      keyword: String?, types: String?,
                      pageSize: Int, page: Int, callback: @escaping POICallback) {
        let req = AMapPOIAroundSearchRequest()
        req.location = AMapGeoPoint.location(withLatitude: CGFloat(latitude), longitude: CGFloat(longitude))
        req.radius = radius; req.keywords = keyword; req.types = types
        req.offset = pageSize; req.page = page
        pendingPOI[ObjectIdentifier(req)] = (req, callback)
        searchAPI?.aMapPOIAroundSearch(req)
    }

    func regeocode(latitude: Double, longitude: Double, callback: @escaping RegeocodeCallback) {
        let req = AMapReGeocodeSearchRequest()
        req.location = AMapGeoPoint.location(withLatitude: CGFloat(latitude), longitude: CGFloat(longitude))
        req.radius = 200; req.requireExtension = true
        pendingRegeo[ObjectIdentifier(req)] = callback
        searchAPI?.aMapReGoecodeSearch(req)
    }

    func geocode(address: String, city: String?, callback: @escaping GeocodeCallback) {
        let req = AMapGeocodeSearchRequest()
        req.address = address; req.city = city
        pendingGeo[ObjectIdentifier(req)] = callback
        searchAPI?.aMapGeocodeSearch(req)
    }

    func drivingRoute(originLat: Double, originLng: Double,
                      destLat: Double, destLng: Double,
                      waypoints: [(lat: Double, lng: Double)],
                      callback: @escaping RouteCallback) {
        let req = AMapDrivingCalRouteSearchRequest()
        req.origin = AMapGeoPoint.location(withLatitude: CGFloat(originLat), longitude: CGFloat(originLng))
        req.destination = AMapGeoPoint.location(withLatitude: CGFloat(destLat), longitude: CGFloat(destLng))
        if !waypoints.isEmpty {
            req.waypoints = waypoints.map {
                AMapGeoPoint.location(withLatitude: CGFloat($0.lat), longitude: CGFloat($0.lng))
            }
        }
        pendingRoute[ObjectIdentifier(req)] = callback
        searchAPI?.aMapDrivingV2RouteSearch(req)
    }

    func walkingRoute(originLat: Double, originLng: Double,
                      destLat: Double, destLng: Double,
                      callback: @escaping RouteCallback) {
        let req = AMapWalkingRouteSearchRequest()
        req.origin = AMapGeoPoint.location(withLatitude: CGFloat(originLat), longitude: CGFloat(originLng))
        req.destination = AMapGeoPoint.location(withLatitude: CGFloat(destLat), longitude: CGFloat(destLng))
        pendingRoute[ObjectIdentifier(req)] = callback
        searchAPI?.aMapWalkingRouteSearch(req)
    }

    func searchDistrict(keywords: String, level: Int, callback: @escaping DistrictCallback) {
        let req = AMapDistrictSearchRequest()
        req.keywords = keywords; req.subdistrict = level
        pendingDistrict[ObjectIdentifier(req)] = callback
        searchAPI?.aMapDistrictSearch(req)
    }

    // MARK: - AMapSearchDelegate

    func onPOISearchDone(_ request: AMapPOISearchBaseRequest!, response: AMapPOISearchResponse!) {
        guard let entry = pendingPOI.removeValue(forKey: ObjectIdentifier(request)) else { return }
        let callback = entry.callback
        let req = entry.request
        guard let response = response else {
            callback(nil, (code: "SEARCH_ERROR", message: "POI search failed"))
            return
        }
        let pois = (response.pois ?? []).map { poi -> POIItem in
            POIItem(
                poiId: poi.uid, title: poi.name, typeDes: poi.type, typeCode: poi.typecode,
                latitude: poi.location.map { Double($0.latitude) },
                longitude: poi.location.map { Double($0.longitude) },
                address: poi.address, tel: poi.tel, distance: Int(poi.distance),
                cityName: poi.city, adName: poi.district
            )
        }
        let pageSize = req.offset > 0 ? Int(req.offset) : 20
        let pageNum  = req.page  > 0 ? Int(req.page)   : 1
        let total    = Int(response.count)
        let pageCount = total > 0 ? (total + pageSize - 1) / pageSize : 0
        callback(POIPage(pois: pois, totalCount: total, pageCount: pageCount, pageNum: pageNum), nil)
    }

    func onReGeocodeSearchDone(_ request: AMapReGeocodeSearchRequest!, response: AMapReGeocodeSearchResponse!) {
        guard let callback = pendingRegeo.removeValue(forKey: ObjectIdentifier(request)) else { return }
        guard let address = response?.regeocode?.formattedAddress else {
            callback(nil, (code: "REGEOCODE_ERROR", message: "Regeocode failed"))
            return
        }
        let ac = response?.regeocode?.addressComponent
        callback(AddressComponent(
            formattedAddress: address, country: ac?.country, province: ac?.province,
            city: ac?.city, cityCode: ac?.citycode, district: ac?.district, adCode: ac?.adcode,
            street: ac?.streetNumber?.street, streetNumber: ac?.streetNumber?.number,
            township: ac?.township, townCode: ac?.towncode
        ), nil)
    }

    func onGeocodeSearchDone(_ request: AMapGeocodeSearchRequest!, response: AMapGeocodeSearchResponse!) {
        guard let callback = pendingGeo.removeValue(forKey: ObjectIdentifier(request)) else { return }
        let items = (response?.geocodes ?? []).map { gc -> GeocodeItem in
            GeocodeItem(
                formattedAddress: gc.formattedAddress, country: gc.country, province: gc.province,
                city: gc.city, cityCode: gc.citycode, district: gc.district, adCode: gc.adcode,
                latitude: gc.location.map { Double($0.latitude) },
                longitude: gc.location.map { Double($0.longitude) },
                level: gc.level
            )
        }
        callback(items, nil)
    }

    func onRouteSearchDone(_ request: AMapRouteSearchBaseRequest!, response: AMapRouteSearchResponse!) {
        guard let callback = pendingRoute.removeValue(forKey: ObjectIdentifier(request)) else { return }
        guard let response = response else {
            callback(nil, (code: "ROUTE_ERROR", message: "Route search failed"))
            return
        }
        let paths = (response.route?.paths ?? []).map { path -> RoutePath in
            let steps = (path.steps ?? []).map { step -> RouteStep in
                RouteStep(
                    instruction: step.instruction, road: step.road,
                    distance: Int(step.distance), duration: Int(step.duration),
                    action: step.action, path: parsePolyline(step.polyline)
                )
            }
            return RoutePath(
                distance: Int(path.distance), duration: Int(path.duration),
                strategy: path.strategy ?? "", tolls: Double(path.tolls),
                tollDistance: Int(path.tollDistance), trafficLights: Int(path.totalTrafficLights),
                steps: steps
            )
        }
        callback(paths, nil)
    }

    func aMapSearchRequest(_ request: Any!, didFailWithError error: Error!) {
        guard let req = request as? AnyObject else { return }
        let key = ObjectIdentifier(req)
        let nsError = error as NSError
        let err = (code: "SEARCH_ERROR_\(nsError.code)", message: nsError.localizedDescription)

        if let entry = pendingPOI.removeValue(forKey: key)      { entry.callback(nil, err); return }
        if let cb    = pendingRegeo.removeValue(forKey: key)     { cb(nil, err); return }
        if let cb    = pendingGeo.removeValue(forKey: key)       { cb(nil, err); return }
        if let cb    = pendingRoute.removeValue(forKey: key)     { cb(nil, err); return }
        if let cb    = pendingDistrict.removeValue(forKey: key)  { cb(nil, err) }
    }

    func onDistrictSearchDone(_ request: AMapDistrictSearchRequest!, response: AMapDistrictSearchResponse!) {
        guard let callback = pendingDistrict.removeValue(forKey: ObjectIdentifier(request)) else { return }
        let items = (response?.districts ?? []).map { convertDistrict($0) }
        callback(items, nil)
    }

    // MARK: - Private helpers

    private func convertDistrict(_ d: AMapDistrict) -> DistrictItem {
        return DistrictItem(
            name: d.name, adCode: d.adcode, cityCode: d.citycode, level: d.level,
            latitude: d.center.map { Double($0.latitude) },
            longitude: d.center.map { Double($0.longitude) },
            districts: (d.districts ?? []).map { convertDistrict($0) }
        )
    }

    private func parsePolyline(_ polyline: String?) -> [[String: Double]] {
        guard let polyline = polyline, !polyline.isEmpty else { return [] }
        return polyline.split(separator: ";").compactMap { seg -> [String: Double]? in
            let parts = seg.split(separator: ",")
            guard parts.count >= 2, let lon = Double(parts[0]), let lat = Double(parts[1]) else { return nil }
            return ["latitude": lat, "longitude": lon]
        }
    }
}

File: packages/autonavi_search_flutter/android/src/main/kotlin/com/autonavi/search/flutter/AMapSearchAdapter.kt
package com.autonavi.search.flutter

import android.content.Context
import com.amap.api.services.core.LatLonPoint
import com.amap.api.services.geocoder.GeocodeAddress
import com.amap.api.services.geocoder.GeocodeQuery
import com.amap.api.services.geocoder.GeocodeResult
import com.amap.api.services.geocoder.GeocodeSearch
import com.amap.api.services.geocoder.RegeocodeQuery
import com.amap.api.services.geocoder.RegeocodeResult
import com.amap.api.services.district.DistrictItem
import com.amap.api.services.district.DistrictResult
import com.amap.api.services.district.DistrictSearch
import com.amap.api.services.district.DistrictSearchQuery
import com.amap.api.services.core.PoiItem
import com.amap.api.services.poisearch.PoiResult
import com.amap.api.services.poisearch.PoiSearch
import com.amap.api.services.route.DrivePath
import com.amap.api.services.route.DriveRouteResult
import com.amap.api.services.route.RouteSearch
import com.amap.api.services.route.WalkPath
import com.amap.api.services.route.WalkRouteResult

/**
 * Adapter that isolates all AMap Search SDK calls.
 * When upgrading the AMap Search SDK, only this file needs changes.
 */
class AMapSearchAdapter(private val context: Context) {

    // MARK: - Plain data types (no SDK types exposed)

    data class POIItem(
        val poiId: String?, val title: String?, val typeDes: String?, val typeCode: String?,
        val latitude: Double?, val longitude: Double?, val address: String?, val tel: String?,
        val distance: Double, val cityName: String?, val adName: String?,
    )

    data class POIPage(val pois: List<POIItem>, val totalCount: Int, val pageCount: Int, val pageNum: Int)

    data class AddressComponent(
        val formattedAddress: String?, val country: String?, val province: String?,
        val city: String?, val cityCode: String?, val district: String?, val adCode: String?,
        val street: String?, val streetNumber: String?, val township: String?, val townCode: String?,
    )

    data class GeocodeItem(
        val formattedAddress: String?, val country: String?, val province: String?,
        val city: String?, val district: String?, val adCode: String?,
        val latitude: Double?, val longitude: Double?, val level: String?,
    )

    data class RouteStep(
        val instruction: String?, val road: String?,
        val distance: Double, val duration: Double,
        val action: String?, val path: List<Map<String, Double>>,
    )

    data class RoutePath(
        val distance: Double, val duration: Double, val strategy: String?,
        val tolls: Double, val tollDistance: Double, val trafficLights: Int,
        val steps: List<RouteStep>,
    )

    data class DistrictNode(
        val name: String?, val adCode: String?, val cityCode: String?, val level: String?,
        val latitude: Double?, val longitude: Double?, val districts: List<DistrictNode>,
    )

    // MARK: - Public API

    fun searchKeyword(
        keyword: String, city: String, types: String,
        page: Int, pageSize: Int,
        callback: (POIPage?, Pair<String, String?>?) -> Unit,
    ) {
        val query = PoiSearch.Query(keyword, types, city).apply {
            pageNum = page - 1; this.pageSize = pageSize
        }
        PoiSearch(context, query).apply {
            setOnPoiSearchListener(object : PoiSearch.OnPoiSearchListener {
                override fun onPoiSearched(result: PoiResult?, code: Int) {
                    if (code == 1000 && result != null) callback(poiResultToPage(result, page), null)
                    else callback(null, "SEARCH_ERROR_$code" to "POI search failed")
                }
                override fun onPoiItemSearched(item: PoiItem?, code: Int) {}
            })
            searchPOIAsyn()
        }
    }

    fun searchNearby(
        lat: Double, lng: Double, radius: Int,
        keyword: String, types: String,
        page: Int, pageSize: Int,
        callback: (POIPage?, Pair<String, String?>?) -> Unit,
    ) {
        val query = PoiSearch.Query(keyword, types, "").apply {
            pageNum = page - 1; this.pageSize = pageSize
        }
        PoiSearch(context, query).apply {
            bound = PoiSearch.SearchBound(LatLonPoint(lat, lng), radius)
            setOnPoiSearchListener(object : PoiSearch.OnPoiSearchListener {
                override fun onPoiSearched(result: PoiResult?, code: Int) {
                    if (code == 1000 && result != null) callback(poiResultToPage(result, page), null)
                    else callback(null, "SEARCH_ERROR_$code" to "Nearby search failed")
                }
                override fun onPoiItemSearched(item: PoiItem?, code: Int) {}
            })
            searchPOIAsyn()
        }
    }

    fun regeocode(lat: Double, lng: Double, callback: (AddressComponent?, Pair<String, String?>?) -> Unit) {
        GeocodeSearch(context).apply {
            setOnGeocodeSearchListener(object : GeocodeSearch.OnGeocodeSearchListener {
                override fun onRegeocodeSearched(result: RegeocodeResult?, code: Int) {
                    if (code == 1000 && result != null) {
                        val addr = result.regeocodeAddress
                        callback(AddressComponent(
                            formattedAddress = addr.formatAddress,
                            country = addr.country, province = addr.province,
                            city = addr.city, cityCode = addr.cityCode,
                            district = addr.district, adCode = addr.adCode,
                            street = addr.streetNumber?.street,
                            streetNumber = addr.streetNumber?.number,
                            township = addr.township, townCode = addr.towncode,
                        ), null)
                    } else {
                        callback(null, "REGEOCODE_ERROR_$code" to "Regeocode failed")
                    }
                }
                override fun onGeocodeSearched(result: GeocodeResult?, code: Int) {}
            })
            getFromLocationAsyn(RegeocodeQuery(LatLonPoint(lat, lng), 200f, GeocodeSearch.AMAP))
        }
    }

    fun geocode(address: String, city: String, callback: (List<GeocodeItem>?, Pair<String, String?>?) -> Unit) {
        GeocodeSearch(context).apply {
            setOnGeocodeSearchListener(object : GeocodeSearch.OnGeocodeSearchListener {
                override fun onRegeocodeSearched(result: RegeocodeResult?, code: Int) {}
                override fun onGeocodeSearched(result: GeocodeResult?, code: Int) {
                    if (code == 1000 && result != null) {
                        val items = result.geocodeAddressList?.map { addr: GeocodeAddress ->
                            GeocodeItem(
                                formattedAddress = addr.formatAddress,
                                country = addr.country, province = addr.province,
                                city = addr.city, district = addr.district,
                                adCode = addr.adcode,
                                latitude = addr.latLonPoint?.latitude,
                                longitude = addr.latLonPoint?.longitude,
                                level = addr.level,
                            )
                        } ?: emptyList()
                        callback(items, null)
                    } else {
                        callback(null, "GEOCODE_ERROR_$code" to "Geocode failed")
                    }
                }
            })
            getFromLocationNameAsyn(GeocodeQuery(address, city))
        }
    }

    fun drivingRoute(
        originLat: Double, originLng: Double,
        destLat: Double, destLng: Double,
        waypoints: List<Pair<Double, Double>>,
        callback: (List<RoutePath>?, Pair<String, String?>?) -> Unit,
    ) {
        val origin = LatLonPoint(originLat, originLng)
        val dest   = LatLonPoint(destLat, destLng)
        val waypointList = waypoints.map { LatLonPoint(it.first, it.second) }
            .takeIf { it.isNotEmpty() }

        val search = RouteSearch(context)
        val fromAndTo = RouteSearch.FromAndTo(origin, dest)
        val query = RouteSearch.DriveRouteQuery(fromAndTo, RouteSearch.DrivingDefault, waypointList, null, "")
        search.setRouteSearchListener(object : RouteSearch.OnRouteSearchListener {
            override fun onDriveRouteSearched(result: DriveRouteResult?, code: Int) {
                if (code == 1000 && result != null) {
                    callback(result.paths.map { drivePathToRoutePath(it) }, null)
                } else {
                    callback(null, "ROUTE_ERROR_$code" to "Driving route failed")
                }
            }
            override fun onBusRouteSearched(p0: com.amap.api.services.route.BusRouteResult?, p1: Int) {}
            override fun onRideRouteSearched(p0: com.amap.api.services.route.RideRouteResult?, p1: Int) {}
            override fun onWalkRouteSearched(p0: WalkRouteResult?, p1: Int) {}
        })
        search.calculateDriveRouteAsyn(query)
    }

    fun walkingRoute(
        originLat: Double, originLng: Double,
        destLat: Double, destLng: Double,
        callback: (List<RoutePath>?, Pair<String, String?>?) -> Unit,
    ) {
        val origin = LatLonPoint(originLat, originLng)
        val dest   = LatLonPoint(destLat, destLng)
        val search = RouteSearch(context)
        val query  = RouteSearch.WalkRouteQuery(RouteSearch.FromAndTo(origin, dest), RouteSearch.WALK_DEFAULT)
        search.setRouteSearchListener(object : RouteSearch.OnRouteSearchListener {
            override fun onDriveRouteSearched(p0: DriveRouteResult?, p1: Int) {}
            override fun onBusRouteSearched(p0: com.amap.api.services.route.BusRouteResult?, p1: Int) {}
            override fun onRideRouteSearched(p0: com.amap.api.services.route.RideRouteResult?, p1: Int) {}
            override fun onWalkRouteSearched(result: WalkRouteResult?, code: Int) {
                if (code == 1000 && result != null) {
                    callback(result.paths.map { walkPathToRoutePath(it) }, null)
                } else {
                    callback(null, "ROUTE_ERROR_$code" to "Walking route failed")
                }
            }
        })
        search.calculateWalkRouteAsyn(query)
    }

    fun searchDistrict(keywords: String, callback: (List<DistrictNode>?, Pair<String, String?>?) -> Unit) {
        DistrictSearch(context).apply {
            query = DistrictSearchQuery().also { it.keywords = keywords }
            setOnDistrictSearchListener(object : DistrictSearch.OnDistrictSearchListener {
                override fun onDistrictSearched(result: DistrictResult?) {
                    val items = result?.district?.firstOrNull()?.subDistrict
                        ?.map { convertDistrict(it) } ?: emptyList()
                    callback(items, null)
                }
            })
            searchDistrictAsyn()
        }
    }

    // MARK: - Private converters

    private fun poiResultToPage(result: PoiResult, page: Int): POIPage {
        val pois = result.pois.map { poi ->
            POIItem(
                poiId = poi.poiId, title = poi.title, typeDes = poi.typeDes,
                typeCode = poi.typeCode,
                latitude  = poi.latLonPoint?.latitude,
                longitude = poi.latLonPoint?.longitude,
                address = poi.snippet, tel = poi.tel,
                distance = poi.distance.toDouble(),
                cityName = poi.cityName, adName = poi.adName,
            )
        }
        val total = (result.query?.pageSize ?: 0) * result.pageCount
        return POIPage(pois = pois, totalCount = total, pageCount = result.pageCount, pageNum = page)
    }

    private fun drivePathToRoutePath(path: DrivePath): RoutePath = RoutePath(
        distance = path.distance.toDouble(), duration = path.duration.toDouble(),
        strategy = path.strategy, tolls = path.tolls.toDouble(),
        tollDistance = path.tollDistance.toDouble(), trafficLights = path.totalTrafficlights,
        steps = path.steps.map { step ->
            RouteStep(
                instruction = step.instruction, road = step.road,
                distance = step.distance.toDouble(), duration = step.duration.toDouble(),
                action = step.action,
                path = step.polyline.map { pt -> mapOf("latitude" to pt.latitude, "longitude" to pt.longitude) },
            )
        },
    )

    private fun walkPathToRoutePath(path: WalkPath): RoutePath = RoutePath(
        distance = path.distance.toDouble(), duration = path.duration.toDouble(),
        strategy = null, tolls = 0.0, tollDistance = 0.0, trafficLights = 0,
        steps = path.steps.map { step ->
            RouteStep(
                instruction = step.instruction, road = step.road,
                distance = step.distance.toDouble(), duration = step.duration.toDouble(),
                action = step.action,
                path = step.polyline.map { pt -> mapOf("latitude" to pt.latitude, "longitude" to pt.longitude) },
            )
        },
    )

    private fun convertDistrict(item: DistrictItem): DistrictNode = DistrictNode(
        name = item.name, adCode = item.adcode, cityCode = item.citycode, level = item.level,
        latitude  = item.center?.latitude,
        longitude = item.center?.longitude,
        districts = item.subDistrict?.map { convertDistrict(it) } ?: emptyList(),
    )
}
